src/websocket/server.cc

```diff
@@ -254,18 +256,23 @@ future<websocket_parser::consumption_result_t> websocket_parser::operator()(
         }
     }
     if (_state == parsing_state::payload) {
-        if (_payload_length > data.size()) {
-            _payload_length -= data.size();
-            remove_mask(data, data.size());
-            _result = std::move(data);
-            return websocket_parser::stop(buff_t(0));
+        if (_buffer.length() + data.size() >= _payload_length) {
+            if (_buffer.length() < _payload_length) {
+                size_t blen = _buffer.length();
+                _buffer.append(data.get(), _payload_length - blen);
+                data.trim_front(_payload_length - blen);
+
+                // We've completed a payload, so reset state and set result.
+                _result = buff_t{_buffer.data(), _buffer.length()};
```

Comment:
i'd suggest
```c++
_result = buff_t{_payload_length};
auto result = std::copy(_buffer.cbegin(), _buffer.cend(), _result.get_write());
_buffer.trim(0);

size_t to_consume = _payload_length - blen;
std::copy_n(data.get(), , to_consume, result);
data.trim_front(_payload_length - blen);

// We've completed a payload, so reset state and set result.
remove_mask(_result, _result.size());
_payload_length = 0;
_state = parsing_state::flags_and_payload_data;
return websocket_parser::stop(std::move(data));
```

for couple reasons:

1. avoid copying the payload twice
2. do not free `_buffer`. we can just reuse it for the next payload
3. name `_payload_length - blen` for better readability